### ISSUE TITLE
Possible new schema for SQLite backend

### ISSUE THREAD

**Issue by plantton** [rformassspectrometry/MsBackendSql#28]
The current schema will use one SQLite table, i.e. `msdata` by default, to store spectra variables including peak values `mz` and `intensity`. For some cases, several `MsBackendSql` objects can share the same underlying `msdata` table, but with different slot index - `rows`. In these cases, the functions which update the values on `msdata` will have a chain of consequences on objects using the same `msdata` table.

Moreover, the current `reset` method will revert the filtering functions on the backends. It might be hard to implement this method on `MsBackendSql`, especially for those filtering functions on peak values. One solution would be store the original peak values on the second temp table. 

Before moving on to the benchmark of this design. I hope I can have some advice on the two table schema: especially whether it's worth to splitting the peak values into the second table?
@lgatto @jorainer

**Comment by jorainer** [comment]
I think splitting the spectra data into a table with just the spectra variables and one with the peak data (i.e. m/z and intensity values) makes total sense. Just be sure to use integer values as primary keys to link between the two - I once realized that MySQL can become slow if primary keys are characters...

Also, I can't remember if in the current implementation the m/z and intensity values are stored as blobs - if so, I would eventually evaluate wheter it would not make more sense (also performance wise) to have the peak data as actual numeric values in the peak table.

For the filter functions: what filter functions on the backend do you mean when you say "especially for those filtering functions on peak values"?

**Comment by lgatto** [comment]
> For the filter functions: what filter functions on the backend do you mean when you say "especially for those filtering functions on peak values"?

`filterMzValues()`, `filterMzRange()` and `filterIntensity()`

**Comment by jorainer** [comment]
These filter functions are only defined and implemented for `Spectra`, not for the backend.

**Comment by plantton** [comment]
> These filter functions are only defined and implemented for `Spectra`, not for the backend.

For `MsBackendSql`, some of the functions are reimplement by SQL statements. For the peak value filtering functions, the `Spectra` implementations will read peak values as `numeric` into R which will cause huge memory cost for large dataset.

I have a new prototype of filtering function on peak values. But this version still keeps peak values as `BLOB` on SQLite backend. The first reason is SQLite lacks of support for array type column. Alternatively, the arrays can be stored as another table in SQLite. But peak values are not always fixed with a certain length. I also tried to use `JSON1` as the solution, but there are other challenges for `JSON` - this format is treated as `characters` by R and SQLite. So the peak values will also cause huge memory issue, and the R packages handle `json` format are also not fast enough to convert `json` back to other format.

**Comment by jorainer** [comment]
Hm, we might need some more thoughts about moving these filter functions then to the backend - or have a way to support that somehow.

For the table with the peak values, I have for example in [CompoundDb](https://github.com/EuracBiomedicalResearch/CompoundDb) the spectra data split into two tables: one for the spectra variables and one for the peak values. The table for the peak values has simply 3 columns, the primary key of the spectrum (in the other table), a column with the m/z values and one with the intensities. No array data type, simply as numbers. That should also simplify to filter on m/z and/or intensity values.

**Comment by lgatto** [comment]
> Hm, we might need some more thoughts about moving these filter functions then to the backend - or have a way to support that somehow.

Not sure what you mean by having a way to support that. But we do need backend-specific implementations to make use of the advantages of the respective backends.

**Comment by jorainer** [comment]
Currently we are caching these operations in the lazy evaluation queue in `Spectra`. If we move these filter functions to the backend we would need to have a lazy evaluation queue also there - something I would want to avoid because it will complicate things.

**Comment by lgatto** [comment]
There are multiple things here:
1. Functions (accessors, filters, ...) need to be able to make use of they advantages of the respective backends.
2. Backends do or don't support writing. When they don't, than lazy is the only viable approach. 
3. When they do, then indeed, there's two options (1) be lazy and process only when the data are accessed or (2) write the data, even if they aren't accessed/returned. We could actually also have write only when returned.

So I agree with you that the lazy evaluation is in Spectra, but at some point, it's the backend that will do the work and write when needed. 

Maybe we should discuss with a concrete example.

**Comment by jorainer** [comment]
I completely agree that for some backends (especially SQL backends) filters on the backend makes sense. I think that's perfectly doable - we could make that depend on whether the filter function is implemented on the backend and then call it immediately or otherwise use the `Spectra`'s caching mechanism. So, I think you can go ahead to implement that filter function for the backend @plantton and I will make sure in the `Spectra` that it will be called.

**Comment by lgatto** [comment]
Thanks! Let's discuss this, as it is a rather complex matter.
